pkg/logs/launchers/file/launcher.go

```diff
@@ -117,10 +123,21 @@ func (s *Launcher) run() {
 		case source := <-s.removedSources:
 			s.removeSource(source)
 		case <-scanTicker.C:
-			s.cleanUpRotatedTailers()
-			// check if there are new files to tail, tailers to stop and tailer to restart because of file rotation
-			s.scan()
+			if s.scanMutex.TryLock() {
+				s.wg.Add(1)
+				go func() {
+					defer s.scanMutex.Unlock()
+					defer s.wg.Done()
+
+					s.cleanUpRotatedTailers()
+					// check if there are new files to tail, tailers to stop and tailer to restart because of file rotation
+					s.scan()
```

Comment:
Instead of granular locking, did you consider pushing 
`files := s.fileProvider.FilesToTail(s.validatePodContainerID, s.activeSources)` to a goroutine instead of the whole scan method? 

AFAIK, `FilesToTail` is the expensive operation, and the majority of `scan` can be done synchronously (to avoid locking). 

`FilesToTail` is mostly stateless, so you could push it's results to a channel, and consume them on the run loop with relatively minimal refactoring. (for example, we may need to copy `s.activeSources`)

Reply:
Okay, that makes sense. The latest commit moves the `FilesToTail` function into it's own go routine and changed the functionality of `addSource` back to it's original behavior. A couple of mutexes are needed however. In order to assure that the `scan` doesn't _remove tailers_ from sources that have gone through `addSource` but have not yet  been processed by `getFiles`, a mutex for the `launcher.addedSourceTailers` was created. Additionally, to ensure that `launcher.activeFiles` wasn't changed while it was being used in `getFiles`, a mutex was added around that.

Reply:
Left a few more comments